In [1]:
import pandas as pd
df=pd.read_csv("messy_sales.csv")

df.shape

(300, 6)

In [2]:
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   order_id  300 non-null    int64  
 1   date      300 non-null    str    
 2   product   300 non-null    str    
 3   price     288 non-null    float64
 4   qty       300 non-null    int64  
 5   zip       300 non-null    int64  
dtypes: float64(1), int64(3), str(2)
memory usage: 19.2 KB


One error is that price only has 288 non-null values with some of them being NaN, that is 12 values unaccounted for. The dates are not in order and have an inconsistent format. Zip code seems to have an error, with 2134 being a 4 digit instead of a typical 5 digit number.

In [4]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [5]:
fill_value = df["price"].median()
df["price"]=df["price"].fillna(fill_value)

df.isna().sum()

order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [6]:
print(df.duplicated().sum())
df = df.drop_duplicates()

df.shape

8


(292, 6)

In [7]:
df["zip"] = df["zip"].astype(str).str.zfill(5)
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,37.53,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,37.53,4,30303
6,1204,04/18/2026,charger,8.41,2,02134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [8]:
df["date"] = pd.to_datetime(df["date"], format="mixed")
df.dtypes

order_id             int64
date        datetime64[us]
product                str
price              float64
qty                  int64
zip                    str
dtype: object

In [9]:
df[df["qty"] <= 0]

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


In [10]:
df=df[df["qty"] > 0]

In [11]:
df.to_csv('sales_clean.csv', index=False)

Cleaning Log

12 columns of prices filled - This is because no value was previously assigned, the median was added to make the data more accurate compared to the mean.

8 Duplicate Results found and dropped - The repeating data are removed because the transaction was likely recorded twice accidentally

Zips become 5 digits - The original file didn't use the correct format for ZIP code so the 0 was added to the beginning to make it 5 digit.

Dates went from str to datetime - The date was treated as a string instead of as a date, the mixed allows the dates to function as is

Negative qtys dropped - The data here is sales data, zero and negative values for quantities are not sales data and should be removed from this list and recorded elsewhere.

Repo Link: https://github.com/KyeSPu/cs82a-portfolio